In [430]:
#importing Libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm

In [431]:
#Importing dataset
df=pd.read_csv("Diabities-210331-154610.csv")

In [432]:
#analyzing Data
df.head(10)

Unnamed: 0  Pregnancies  Glucose  blood pressure  skin thickness  Insulin  \
0         NaN          6.0    148.0            72.0            35.0      0.0   
1         NaN          1.0     85.0            66.0            29.0      0.0   
2         NaN          8.0    183.0            64.0             0.0      0.0   
3         NaN          1.0     89.0            66.0            23.0     94.0   
4         NaN          0.0    137.0            40.0            35.0    168.0   
5         NaN          5.0    116.0            74.0             0.0      0.0   
6         NaN          3.0     78.0            50.0            32.0     88.0   
7         NaN         10.0    115.0             0.0             0.0      0.0   
8         NaN          2.0    197.0            70.0            45.0    543.0   
9         NaN          8.0    125.0            96.0             0.0      0.0   

    BMI  DiabetesPedigreeFunction   Age  Outcome  
0  33.6                     0.627  50.0      1.0  
1  26.6                     0.351  31.0      0.0  
2  23.3                     0.672  32.0      1.0  
3  28.1                     0.167  21.0      0.0  
4  43.1                     2.288  33.0      1.0  
5  25.6                     0.201  30.0      0.0  
6  31.0                     0.248  26.0      1.0  
7  35.3                     0.134  29.0      0.0  
8  30.5                     0.158  53.0      1.0  
9   0.0                     0.232  54.0      1.0

In [433]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 769 entries, 0 to 768
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                0 non-null      float64
 1   Pregnancies               768 non-null    float64
 2   Glucose                   768 non-null    float64
 3   blood pressure            768 non-null    float64
 4   skin thickness            768 non-null    float64
 5   Insulin                   768 non-null    float64
 6   BMI                       768 non-null    float64
 7   DiabetesPedigreeFunction  768 non-null    float64
 8   Age                       768 non-null    float64
 9   Outcome                   768 non-null    float64
dtypes: float64(10)
memory usage: 60.2 KB


In [434]:
df.shape

(769, 10)

In [435]:
df['Outcome'].value_counts()

0.0    500
1.0    268
Name: Outcome, dtype: int64

In [436]:
df.groupby('Outcome').mean()

Unnamed: 0  Pregnancies     Glucose  blood pressure  skin thickness  \
Outcome                                                                        
0.0             NaN     3.298000  109.980000       68.184000       19.664000   
1.0             NaN     4.865672  141.257463       70.824627       22.164179   

            Insulin        BMI  DiabetesPedigreeFunction        Age  
Outcome                                                              
0.0       68.792000  30.304200                  0.429734  31.190000  
1.0      100.335821  35.142537                  0.550500  37.067164

In [437]:
#data cleaning
df.isnull().sum()

Unnamed: 0                  769
Pregnancies                   1
Glucose                       1
blood pressure                1
skin thickness                1
Insulin                       1
BMI                           1
DiabetesPedigreeFunction      1
Age                           1
Outcome                       1
dtype: int64

In [438]:
df = df.dropna(how='all')

In [439]:
np.any(np.isnan(df))
df.fillna(999, inplace=True)
df = df.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)

In [440]:
X=df.iloc[:,:-1]
y=df.iloc[:,-1]

In [441]:
print(X)

     Unnamed: 0  Pregnancies  Glucose  blood pressure  skin thickness  \
0         999.0          6.0    148.0            72.0            35.0   
1         999.0          1.0     85.0            66.0            29.0   
2         999.0          8.0    183.0            64.0             0.0   
3         999.0          1.0     89.0            66.0            23.0   
4         999.0          0.0    137.0            40.0            35.0   
..          ...          ...      ...             ...             ...   
763       999.0         10.0    101.0            76.0            48.0   
764       999.0          2.0    122.0            70.0            27.0   
765       999.0          5.0    121.0            72.0            23.0   
766       999.0          1.0    126.0            60.0             0.0   
767       999.0          1.0     93.0            70.0            31.0   

     Insulin   BMI  DiabetesPedigreeFunction   Age  
0        0.0  33.6                     0.627  50.0  
1        0.0  26.

In [442]:
print(y)

0      1.0
1      0.0
2      1.0
3      0.0
4      1.0
      ... 
763    0.0
764    0.0
765    0.0
766    1.0
767    0.0
Name: Outcome, Length: 768, dtype: float64


In [450]:
#Splitting Data
X_train, X_test, y_train , y_test=train_test_split(X,y,test_size=25,random_state=0)

In [451]:
print(X.shape,X_train.shape,X_test.shape)

(768, 9) (743, 9) (25, 9)


In [452]:
#CLASSIFICATION

### RANDOM FOREST CLASSIFICATION

In [453]:
%timeit

In [468]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier(n_estimators=100, oob_score=True,criterion='entropy',n_jobs=1,random_state=1)
classifier.fit(X_train,y_train)
y_pred=classifier.predict(X_test)

In [469]:
from sklearn.metrics import accuracy_score
accuracy=accuracy_score(y_pred,y_test)*100

In [470]:
print("Accuracy using random forest classifier is", accuracy)

Accuracy using random forest classifier is 92.0


### LOGISTIC REGRESSION

In [471]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,r2_score,classification_report
LR=LogisticRegression(solver='lbfgs',max_iter=1000)
LR.fit(X_train,y_train)
y_pred=LR.predict(X_test)

In [472]:
accuracy1=accuracy_score(y_pred,y_test)*100

In [473]:
print("Accuracy using Logistic Regression is",accuracy1)

Accuracy using Logistic Regression is 96.0


### K NEAREST NEIGHBOUR

In [474]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train,y_train)
y_pred=knn.predict(X_test)

In [475]:
accuracy2=accuracy_score(y_pred,y_test)*100
print("Accuracy using k Nearest Neighbors is", accuracy2)

Accuracy using k Nearest Neighbors is 88.0


### SUPPORT VECTOR MACHINE

In [476]:
classifier=svm.SVC(kernel='linear')

In [477]:
classifier.fit(X_train,y_train)

SVC(kernel='linear')

In [478]:
X_train_prediction=classifier.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,y_train)*100

In [479]:
print("Accuracy score of Training data",training_data_accuracy)

Accuracy score of Training data 76.98519515477793


In [480]:
X_test_prediction=classifier.predict(X_test)
test_data_accuracy=accuracy_score(X_test_prediction,y_test)*100

In [481]:
print("Accuracy score of Test data",test_data_accuracy)

Accuracy score of Test data 92.0


### The data is classified using four different classifiers 
#### 1) Random Forest Classifier with 92% accuracy
#### 2) Logistic Regression with 96% accuracy
#### 3) KNN with 88% accuracy
#### 4) SVM with 92% accuracy
#### Since Logistic Regression has the highest accuracy therefore it is the most suitable model for classifying the data for patients with diabetes.